kernel/aster-nix/src/syscall/preadv.rs

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: MPL-2.0
+use super::SyscallReturn;
+use crate::{
+    fs::file_table::FileDesc,
+    prelude::*,
+    util::{read_val_from_user, write_bytes_to_user},
+};
+
+pub fn sys_readv(fd: FileDesc, io_vec_ptr: Vaddr, io_vec_count: usize) -> Result<SyscallReturn> {
+    let res = do_sys_readv(fd, io_vec_ptr, io_vec_count)?;
+    Ok(SyscallReturn::Return(res as _))
+}
+
+pub fn sys_preadv(
+    fd: FileDesc,
+    io_vec_ptr: Vaddr,
+    io_vec_count: usize,
+    offset: i64,
+) -> Result<SyscallReturn> {
+    let res = do_sys_preadv(fd, io_vec_ptr, io_vec_count, offset, RWFFlag::empty())?;
+    Ok(SyscallReturn::Return(res as _))
+}
+
+pub fn sys_preadv2(
+    fd: FileDesc,
+    io_vec_ptr: Vaddr,
+    io_vec_count: usize,
+    offset: i64,
+    flags: u32,
+) -> Result<SyscallReturn> {
+    let flags = match RWFFlag::from_bits(flags) {
+        Some(flags) => flags,
+        None => return_errno_with_message!(Errno::EINVAL, "invalid flags"),
+    };
+    let res = if offset == -1 {
+        do_sys_readv(fd, io_vec_ptr, io_vec_count)?
+    } else {
+        do_sys_preadv(fd, io_vec_ptr, io_vec_count, offset, flags)?
+    };
+    Ok(SyscallReturn::Return(res as _))
+}
+
+fn do_sys_preadv(
+    fd: FileDesc,
+    io_vec_ptr: Vaddr,
+    io_vec_count: usize,
+    offset: i64,
+    _flags: RWFFlag,
+) -> Result<usize> {
+    debug!(
+        "preadv: fd = {}, io_vec_ptr = 0x{:x}, io_vec_counter = 0x{:x}, offset = 0x{:x}",
+        fd, io_vec_ptr, io_vec_count, offset
+    );
+
+    if offset < 0 {
+        return_errno_with_message!(Errno::EINVAL, "offset cannot be negative");
+    }
+
+    let file = {
+        let current = current!();
+        let filetable = current.file_table().lock();
+        filetable.get_file(fd)?.clone()
+    };
+
+    if io_vec_count == 0 {
+        return Ok(0);
+    }
+
+    // Calculate the total buffer length and check for overflow
+    let total_len = io_vec_count
```

Comment:
The variable `total_len` does not mean the total buffer length, but the end address of io_vecs, right? A more suitable variable name should be used and revise the docs.

Reply:
Maybe you want check the following error in [the manual page](https://man7.org/linux/man-pages/man2/readv.2.html).

```plain
EINVAL The sum of the iov_len values overflows an ssize_t value.
```
The error is not covered now.

But it should be done by checking the sum of `io_vec.len()`, insead of `io_vec_count * core::mem::size_of::<IoVec>()`

Reply:
Thank you for your consideration! I have replaced it in the newest version now.